### PR TITLE
Build rocksdb and arm images in amd64 hosts (ubuntu 20.04 hosts)

### PR DESCRIPTION
**Pull request by darobs**
When this project was started, we used base images to run some commands in the native execution environment (for example: create a user in the container).  Now with cross-platform docker builds, we can use runtime emulation to do this work, and not maintain the base images.

We also had an inconsistent way to house the RocksDB library files.  This PR builds the libraries in parallel with the project executables, and collects the libraries with the other artifacts for the image build to use. I also updated the container images build to ubuntu 20.04 to remove all references to qemu-static container.

This PR reduces (but does not completely eliminate) dependency on docker hub for the project.

Testing for this PR was done through executing E2E tests with images built from this PR.

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [x] concise summary of tests added/modified
	- [x] local testing done.

### CODE REVIEW

File: builds/misc/templates/build-images.yaml [hunk 1]
```diff
@@ -27,6 +27,43 @@ stages:
...
             name: check_files
 
+################################################################################
+  - stage: BuildRocksDB
+################################################################################
+    displayName: Build RocksDB
+    pool:
+      name: $(pool.linux.name)
+      demands:
+        - ImageOverride -equals agent-aziotedge-ubuntu-18.04-docker
```
[the remark this answers is not in the document]
Comment (gauravIoTEdge) [reply builds/misc/templates/build-images.yaml:37]: This actually steps on something that I'd been wondering about: Is there a better way of "storing" these intermediate stages/steps?

It seems to me that we duplicate a lot of code just by Ctrl+C/Ctrl+V the same steps, and end up with a lot of pipelines that are mostly similar, but not quite the same.

Not to overreact for the same 18.04/20.04 miss, but this is one of the annoyances we could avoid by maybe splitting recurring stages into their own YAML files?

(Not suggesting that needs to happen as a part of this work, but we have to start brainstorming _somewhere_)

File: edge-agent/docker/linux/amd64/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,4 @@
 ARG base_tag=3.1.19-alpine3.14
```
Comment (gauravIoTEdge) [inline edge-agent/docker/linux/amd64/Dockerfile:1]: This is orthogonal to the work at hand, and it makes sense to do it as a part of a separate check-in. That said, it doesn't hurt to mention:

IIRC @damonbarry wanted to remote the base_tag from all these Dockerfiles. If you're touching them.....

File: edge-hub/docker/linux/arm32v7/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,21 @@
-ARG base_tag=1.0.6.15-linux-arm32v7
-FROM azureiotedge/azureiotedge-hub-base:${base_tag}
+ARG base_tag=3.1.19-bionic-arm32v7
+FROM mcr.microsoft.com/dotnet/aspnet:${base_tag}
+
+# Add an unprivileged user account for running Edge Hub
+ARG EDGEHUBUSER_ID=13623
```
Comment (gauravIoTEdge) [inline edge-hub/docker/linux/arm32v7/Dockerfile:5]: Is 13623 a randomly picked number or is it magic?

Please consider elaborating in comment.

File: test/modules/EdgeHubRestartTester/docker/linux/arm64v8/Dockerfile [hunk 4]
```diff
@@ -1,5 +1,9 @@
-ARG base_tag=1.0.6.15-linux-arm64v8
-FROM azureiotedge/azureiotedge-module-base:${base_tag}
+ARG base_tag=3.1.19-bionic-arm64v8
```
Comment (gauravIoTEdge) [inline test/modules/EdgeHubRestartTester/docker/linux/arm64v8/Dockerfile:1]: Eventually, we will need to update these tags. 

We really should move to env_file: https://docs.docker.com/compose/compose-file/compose-file-v3/#env_file

$ docker run --env-file=env_file_name ... .... ?

Should work..?

Comment (gauravIoTEdge) [reply test/modules/EdgeHubRestartTester/docker/linux/arm64v8/Dockerfile:1]: Really, this is the kind of thing that could be automated -- most of these Dockerfiles are very similar, with like 2 lines that are different (module/DLL name). 

If nothing else, we could have per-module templates? Cost-Benefit conversation will need to be had, of course.

**Review by gauravIoTEdge — APPROVED** [review]
We really should have a better way to auto-populate the $(version), rather than have it be entered manually. For something like this, we end up touching the build as well as release images pipelines anyway.

File: builds/misc/images-release.yaml [hunk 5]
```diff
@@ -305,6 +305,43 @@ stages:
...
     - template: templates/build-broker-watchdog.yaml
 
+################################################################################
+  - stage: BuildRocksDB
+################################################################################
+    displayName: Build RocksDB
+    pool:
+      name: $(pool.linux.name)
+      demands:
+        - ImageOverride -equals agent-aziotedge-ubuntu-18.04-docker
```
[the remark this answers is not in the document]
Comment (darobs) [reply builds/misc/images-release.yaml:315]: I limited the scope to changes to the stages which required 20.04.  I'm totally OK with moving all stages to 20.04, I'm just being cautious.

File: edge-hub/docker/linux/arm32v7/Dockerfile (hunk 3 above)
Comment (darobs) [reply edge-hub/docker/linux/arm32v7/Dockerfile:5]: https://github.com/Azure/iotedge/pull/4998

The original problem was that the default user was 1000 - that typically corresponds to the first user created on a Linux system, and may not be a user which should have R/W permission on edgeAgent/edgeHub files. The proposed fix was to pick a semi-arbitrary number above 10000 to avoid typical regular users, and 13622 and 13623 were chosen for edgeAgent and edgeHub, respectively.  I can elaborate this comment.

File: builds/misc/templates/build-images.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (darobs) [reply builds/misc/templates/build-images.yaml]: I did separate the "build RocksDB" stage out to its own file - which is something I should have done earlier.

**Review by gauravIoTEdge — APPROVED** [review]

File: edge-agent/docker/linux/arm64v8/Dockerfile [hunk 6]
```diff
@@ -1,6 +1,26 @@
-ARG base_tag=1.0.6.15-linux-arm64v8
+ARG base_tag=3.1.19-bionic-arm64v8
+ARG num_procs=4
```
Comment (nimanch) [inline edge-agent/docker/linux/arm64v8/Dockerfile:2]: why was this added?

File: edge-hub/docker/linux/arm32v7/Dockerfile (hunk 3 above)
Comment (nimanch) [reply edge-hub/docker/linux/arm32v7/Dockerfile:6]: @darobs 
I thought the hubStart.sh script executed as part of the container already does that?

File: edge-agent/docker/linux/arm64v8/Dockerfile (hunk 6 above)
Comment (darobs) [reply edge-agent/docker/linux/arm64v8/Dockerfile:2]: @nimanch It was in the original base image: See https://github.com/Azure/iotedge/blob/release/1.2/edge-agent/docker/linux/arm64v8/base/Dockerfile

...but it doesn't need to be there.
